src/python/pants/pantsd/pants_daemon.py

```diff
@@ -502,7 +515,8 @@ def needs_restart(self, option_fingerprint):
     """
     should_shutdown_after_run = self._bootstrap_options.for_global_scope().shutdown_pantsd_after_run
     return super().needs_restart(option_fingerprint) or \
-           (self.is_alive() and should_shutdown_after_run)
+           (self.is_alive() and should_shutdown_after_run) or \
+           self._has_exceeded_memory_usage()
```

Comment:
Thanks @blorente! This is great.

Given there are so many reasons that a daemon can restart now, I'd advocate explicitly `INFO` the restart event when it does. Is it somewhere around https://github.com/pantsbuild/pants/blob/14f079f03bc0fa792241d80944e890c844efedce/src/python/pants/java/nailgun_executor.py#L169-L187?

Reply:
I have experimented with some designs for this. The main impedinent for `[INFO]`ing this is that we want it to end up in the pantsd log (as opposed to the client console output). However, by the time we can log to the pantsd log, we have lost the information we needed to record the cause (because the causes get calculated in the client process). The best solution I can think of is to record the causes somewhere (a file or a global variable) to be able to traverse the process boundary, but I really don't like that.

However, I realized that line 518 is not needed anymore, so I removed it, and now we have less causes of death. I have also added some logging in the pantsd log side to mark when we are restarting because of the memory flag.

I think this will be enough, but would love your thoughts.

Reply:
> by the time we can log to the pantsd log, we have lost the information we needed to record the cause (because the causes get calculated in the client process).

Yes I understand this might be difficult. I think as long as user can get a restart message (doesn't have to include a reason), that should be enough, as it appears right now, the restart message only shows up in pantsd log.

This looks good, so not blocking.

Reply:
Ohh, I understand now :) I agree that something could be displayed to the user log. Added that now. Thanks for the input!